src/runtime/server.ts

```diff
@@ -1,103 +1,187 @@
+			},
+			status: 404,
+		},
+	);
+}
+
+function badBodyResponse(): Response {
+	return new Response(
+		JSON.stringify({
+			"message": "Request must have a valid JSON body.",
+		}),
+		{
+			headers: {
+				"Content-Type": "application/json",
+				"Access-Control-Allow-Origin": "*",
+			},
+			status: 400,
+		},
+	);
+}
+
+function serverError(e: unknown): Response {
+	const status = e instanceof RuntimeError ? e.statusCode : 500;
+	return new Response(
+		JSON.stringify({
+			"message": "Request must have a valid JSON body.",
+		}),
+		{
+			headers: {
+				"Content-Type": "application/json",
+				"Access-Control-Allow-Origin": "*",
+			},
+			status,
+		},
+	);
+}
+
+async function handleScriptCall<DependenciesSnakeT, DependenciesCamelT>(
+	runtime: Runtime<DependenciesSnakeT, DependenciesCamelT>,
+	url: URL,
+	moduleName: string,
+	scriptName: string,
+	script: Script,
+	req: Request,
+	info: Deno.ServeHandlerInfo,
+) {
+	// If a script is not public, return 404
+	if (!script.public) return notFoundResponse();
+
+	// Create context
+	const ctx = runtime.createRootContext({
+		httpRequest: {
+			method: req.method,
+			path: url.pathname,
+			remoteAddress: info.remoteAddr.hostname,
+			headers: Object.fromEntries(req.headers.entries()),
+		},
+	});
+
+	// Parse body
+	let body;
+	try {
+		body = await req.json();
+	} catch {
+		return badBodyResponse();
+	}
+
+	let output: any;
+	try {
+		// Call module
+		output = await ctx.call(
+			moduleName as any,
+			scriptName as any,
+			body,
+		);
+	} catch (e) {
+		return serverError(e);
+	}
+
+	if (output?.__tempPleaseSeeOGBE3_NoData) return new Response(
+		undefined,
+		{
+			status: 204,
+			headers: {
+				"Access-Control-Allow-Origin": "*",
+			},
+		},
+	);
+	else return new Response(
+		JSON.stringify(output),
+		{
+			status: 200,
+			headers: {
+				"Content-Type": "application/json",
+				"Access-Control-Allow-Origin": "*",
+			},
+		},
+	);
+}
+
+async function handleApiRouteCall<DependenciesSnakeT, DependenciesCamelT>(
+	runtime: Runtime<DependenciesSnakeT, DependenciesCamelT>,
+	url: URL,
+	moduleName: string,
+	routeName: string,
+	apiRoute: ApiRoute,
+	req: Request,
+	info: Deno.ServeHandlerInfo,
+) {
+	if (!apiRoute.methods.has(req.method)) notFoundResponse();
+
+	// Create context
+	const ctx = runtime.createRootApiContext(
+		{
+			httpRequest: {
+				method: req.method,
+				path: url.pathname,
+				remoteAddress: info.remoteAddr.hostname,
+				headers: Object.fromEntries(req.headers.entries()),
+			},
+		},
+		moduleName,
+		routeName,
+	);
+
+	// Call route
+	const res = await ctx.runBlock(async () => await apiRoute.run(ctx, req));
+	console.log(
+		`API Response ${moduleName}.${routeName}:\n${JSON.stringify(res, null, 2)}`,
+	);
+
+	return res;
+}
+
 export function serverHandler<DependenciesSnakeT, DependenciesCamelT>(
 	runtime: Runtime<DependenciesSnakeT, DependenciesCamelT>,
 ): Deno.ServeHandler {
+	const resolver = new PathResolver(runtime.apiPaths());
```

Comment:
can we implement this so there's no intermediary state? just to serve a single request in a cf worker, you need to build the whole array for `apiPaths` then assemble it with `PathResolver`.

we should generate data in to the runtime config in the final form used for fast lookups. there should be little to no intermediary calls for handling a request.

afaik the fastest way is to build a tree for the routes.

Reply:
Sounds good— however, no matter what, serving a single request will  STILL be `O(n + k)` (with `n` being the number of API routes and `k` being the number of route segments)

Either way, you need to load a runtime, which knows about every module, every script, and *every* API route.

In the end, no matter how it's implemented, whether it's a separate class or not, `PathResolver` will be at least `O(k)` for resolving a path and at least `O(n)` for building.

Reply:
Rerequesting review for now— can discuss more if you want.

Reply:
you could pre-build the routes as a tree and pass that in to the runtime. i'd rather get this shipped faster, though. if you wanted to go overboard, you could even implement the entire router as a tree, including the routes for all of the module scripts.

i want to clarify – this runs in a serverless context, so the time it takes to assemble `pathresolver` impacts the request time for every request.

Reply:
Will make a ticket about speed and do more investigation.